### ISSUE TITLE
Wrong way of throwing exception + problem with -r option in backsmali

### ISSUE THREAD

**Issue by stiv-yakovenko** [JesusFreke/smali#482]
This always gives java.util.MissingFormatArgumentException: Format specifier '%s':

> if (entryNames.size() == 0) {
>       throw new NoDexException("%s contains no dex file");
>  }

because 2nd parameter is missing. If I put breakpoint here, i see in debuger that container is ZipDexContainer with reference to c:\tmp\system\framework\core-libart.jar. 

I start backsmali this way:

d -a 23 -r ALL ./tmp/split.dex -d c:/tmp/system/framework -o ./testdex/cmdline/testout

I've downloaded entire folder c:/tmp/system/framework from my android emulator.
Without -r all works ok.
What can I do to finally disassemble my DEX with registers info?

**Comment by JesusFreke** [comment]
Yep, looks like a bug. I'll get it fixed :)

As a workaround, I think this should work:

baksmali d -a 23 -r ALL ./tmp/split.dex -b c:/tmp/system/framework/arm/boot.oat -o ./testdex/cmdline/testout

Thanks for the report!

**Comment by stiv-yakovenko** [comment]
I've found out that -b "" also helps. Please explain why (by default) do you need all these dependencies for register types derivation?
Thanx

**Comment by JesusFreke** [comment]
It needs to have the full classpath available, because it needs to know about the full inheritance hierarchy of any classes that are being used.

For example, consider something like

```
(pseudocode)
v0 = random integer
if (v0 > 0) {
  v1 = new Class1
} else {
  v1 = new Class2
}
```
After this block of code, the register type of v1 depends on the relationship between Class1 and Class2. If they are completely unrelated, then the type of v1 will be java.lang.Object. Or if Class2 is a subclass of Class1 then the type of v1 will be Class1. Or if Class1 and Class2 share some common superclass, say Class0, then the type of v1 will be Class0.

The only way that this type of analysis can be performed is if we have full information about all classes on the classpath.

**Comment by JesusFreke** [comment]
The missing format argument is fixed in 152250d90068f942b59f88465f198e0f75acdf8e
